### PR TITLE
feat: complete invocation stub

### PR DESCRIPTION
**Pull request by smoyer64**
This PR includes the code to create, seal and unseal `invocation` `Token`s according to the specification at https://github.com/ucan-wg/invocation.  "Sealed" `invocation` tokens have the form (shown in DAG-JSON for humans):
```
[
  {
    "/": {
      "bytes": "o/vTvTs8SEkD9QL/eNhhW0fAng/SGBouywCbUnOfsF2RFHxaV02KTCyzgDxlJLZ2XN/Vk5igLmlKL3QIXMaeCQ"
    }
  },
  {
    "h": {
      "/": {
        "bytes": "NO0BcQ"
      }
    },
    "ucan/inv@1.0.0-rc.1": {
      "args": {
        "headers": {
          "Content-Type": "application/json"
        },
        "payload": {
          "body": "UCAN is great",
          "draft": true,
          "title": "UCAN for Fun and Profit",
          "topics": [
            "authz",
            "journal"
          ]
        },
        "uri": "https://example.com/blog/posts"
      },
      "cmd": "/crud/create",
      "exp": 1730812145,
      "iss": "did:key:z6MkvMGkN5nbUQLBVqJhr13Zdqyh9rR1VuF16PuZbfocBxpv",
      "meta": {
        "env": "development",
        "tags": [
          "blog",
          "post",
          "pr#123"
        ]
      },
      "nonce": {
        "/": {
          "bytes": "q1AH6MJrqoTH6av7"
        }
      },
      "prf": [
        {
          "/": "bafyreigx3qxd2cndpe66j2mdssj773ecv7tqd7wovcnz5raguw6lj7sjoe"
        },
        {
          "/": "bafyreib34ira254zdqgehz6f2bhwme2ja2re3ltcalejv4x4tkcveujvpa"
        },
        {
          "/": "bafyreibkb66tpo2ixqx3fe5hmekkbuasrod6olt5bwm5u5pi726mduuwlq"
        }
      ],
      "sub": "did:key:z6MkuFj35aiTL7YQiVMobuSeUQju92g7wZzufS3HAc6NFFcQ"
    }
  }
]
```

The API and documentation for this package is shown below - please poke holes in both!

```
package invocation // import "github.com/ucan-wg/go-ucan/token/invocation"

Package invocation implements the UCAN invocation specification with an
immutable Token type as well as methods to convert the Token to and from the
envelope-enclosed, signed and DAG-CBOR-encoded form that should most commonly be
used for transport and storage.

[envelope]: https://github.com/ucan-wg/spec#envelope
[invocation]: https://github.com/ucan-wg/invocation

CONSTANTS

const Tag = "ucan/inv@1.0.0-rc.1"
    Tag is the string used as a key within the SigPayload that identifies that
    the TokenPayload is an invocation.

[Tag]: https://github.com/ucan-wg/invocation#type-tag


TYPES

type Option func(*Token) error
    Option is a type that allows optional fields to be set during the creation
    of a invocation Token.

func WithArgument(key string, val datamodel.Node) Option
    WithArgument adds a key/value pair to the Token's Arguments field.

func WithArguments(args map[string]datamodel.Node) Option
    WithArguments sets the Token's Arguments field to the provided map.

    Note that this will overwrite any existing Arguments whether provided by a
    previous call to this function or by one or more calls to WithArgument.

func WithAudience(aud did.DID) Option
    WithAudience sets the Token's audience to the provided did.DID.

    If the provided did.DID is the same as the Token's subject, the audience is
    not set.

func WithCause(cause *cid.Cid) Option
    WithCause sets the Token's cause field to the provided cid.Cid.

func WithEmptyNonce() Option
    WithEmptyNonce sets the Token's nonce to an empty byte slice as suggested by
    the UCAN spec for invocation tokens that represent idempotent operations.

func WithExpiration(exp time.Time) Option
    WithExpiration set's the Token's optional "expiration" field to the value of
    the provided time.Time.

func WithExpirationIn(after time.Duration) Option
    WithExpirationIn set's the Token's optional "expiration" field to Now() plus
    the given duration.

func WithInvokedAt(iat time.Time) Option
    WithInvokedAt sets the Token's invokedAt field to the provided time.Time.

    If this Option is not provided, the invocation Token's iat field will be
    set to the value of time.Now(). If you want to create an invocation Token
    without this field being set, use the WithoutInvokedAt Option.

func WithInvokedAtIn(after time.Duration) Option
    WithInvokedAtIn sets the Token's invokedAt field to Now() plus the given
    duration.

func WithMeta(key string, val any) Option
    WithMeta adds a key/value pair in the "meta" field.

    WithMeta can be used multiple times in the same call. Accepted types for the
    value are: bool, string, int, int32, int64, []byte, and ipld.Node.

func WithNonce(nonce []byte) Option
    WithNonce sets the Token's nonce with the given value.

    If this option is not used, a random 12-byte nonce is generated for this
    required field. If you truly want to create an invocation Token without a
    nonce, use the WithEmptyNonce Option which will set the nonce to an empty
    byte array.

func WithoutInvokedAt() Option
    WithoutInvokedAt clears the Token's invokedAt field.

type Token struct {
        // Has unexported fields.
}
    Token is an immutable type that holds the fields of a UCAN invocation.

func Decode(b []byte, decFn codec.Decoder) (*Token, error)
    Decode unmarshals the input data using the format specified by the provided
    codec.Decoder into a Token.

    An error is returned if the conversion fails, or if the resulting Token is
    invalid.

func DecodeReader(r io.Reader, decFn codec.Decoder) (*Token, error)
    DecodeReader is the same as Decode, but accept an io.Reader.

func FromDagCbor(data []byte) (*Token, error)
    FromDagCbor unmarshals the input data into a Token.

    An error is returned if the conversion fails, or if the resulting Token is
    invalid.

func FromDagCborReader(r io.Reader) (*Token, error)
    FromDagCborReader is the same as FromDagCbor, but accept an io.Reader.

func FromDagJson(data []byte) (*Token, error)
    FromDagJson unmarshals the input data into a Token.

    An error is returned if the conversion fails, or if the resulting Token is
    invalid.

func FromDagJsonReader(r io.Reader) (*Token, error)
    FromDagJsonReader is the same as FromDagJson, but accept an io.Reader.

func FromIPLD(node datamodel.Node) (*Token, error)
    FromIPLD decode the given IPLD representation into a Token.

func FromSealed(data []byte) (*Token, cid.Cid, error)
    FromSealed decodes the provided binary data from the DAG-CBOR format,
    verifies that the envelope's signature is correct based on the public key
    taken from the issuer (iss) field and calculates the CID of the incoming
    data.

func FromSealedReader(r io.Reader) (*Token, cid.Cid, error)
    FromSealedReader is the same as Unseal but accepts an io.Reader.

func New(iss, sub did.DID, cmd command.Command, prf []cid.Cid, opts ...Option) (*Token, error)
    New creates an invocation Token with the provided options.

    If no nonce is provided, a random 12-byte nonce is generated. Use the
    WithNonce or WithEmptyNonce options to specify provide your own nonce or to
    leave the nonce empty respectively.

    If no invokedAt is provided, the current time is used. Use the
    WithInvokedAt or WithInvokedAtIn Options to specify a different time or the
    WithoutInvokedAt Option to clear the Token's invokedAt field.

    With the exception of the WithMeta option, all other will overwrite the
    previous contents of their target field.

func (t *Token) Arguments() map[string]datamodel.Node
    Arguments returns the arguments to be used when the command is invoked.

func (t *Token) Audience() did.DID
    Audience returns the did.DID representing the Token's audience.

func (t *Token) Cause() *cid.Cid
    Cause returns the Token's (optional) cause field which may specify which
    describes the Receipt that requested the invocation.

func (t *Token) Command() command.Command
    Command returns the capability's command.Command.

func (t *Token) Encode(privKey crypto.PrivKey, encFn codec.Encoder) ([]byte, error)
    Encode marshals a Token to the format specified by the provided
    codec.Encoder.

func (t *Token) EncodeWriter(w io.Writer, privKey crypto.PrivKey, encFn codec.Encoder) error
    EncodeWriter is the same as Encode, but accepts an io.Writer.

func (t *Token) Expiration() *time.Time
    Expiration returns the time at which the Token expires.

func (t *Token) InvokedAt() *time.Time
    InvokedAt returns the time.Time at which the invocation token was created.

func (t *Token) Issuer() did.DID
    Issuer returns the did.DID representing the Token's issuer.

func (t *Token) Meta() *meta.Meta
    Meta returns the Token's metadata.

func (t *Token) Nonce() []byte
    Nonce returns the random Nonce encapsulated in this Token.

func (t *Token) Proof() []cid.Cid
    Proof() returns the ordered list of cid.Cids which referenced the delegation
    Tokens that authorize this invocation.

func (t *Token) Subject() did.DID
    Subject returns the did.DID representing the Token's subject.

func (t *Token) ToDagCbor(privKey crypto.PrivKey) ([]byte, error)
    ToDagCbor marshals the Token to the DAG-CBOR format.

func (t *Token) ToDagCborWriter(w io.Writer, privKey crypto.PrivKey) error
    ToDagCborWriter is the same as ToDagCbor, but it accepts an io.Writer.

func (t *Token) ToDagJson(privKey crypto.PrivKey) ([]byte, error)
    ToDagJson marshals the Token to the DAG-JSON format.

func (t *Token) ToDagJsonWriter(w io.Writer, privKey crypto.PrivKey) error
    ToDagJsonWriter is the same as ToDagJson, but it accepts an io.Writer.

func (t *Token) ToSealed(privKey crypto.PrivKey) ([]byte, cid.Cid, error)
    ToSealed wraps the invocation token in an envelope, generates the signature,
    encodes the result to DAG-CBOR and calculates the CID of the resulting
    binary data.

func (t *Token) ToSealedWriter(w io.Writer, privKey crypto.PrivKey) (cid.Cid, error)
    ToSealedWriter is the same as ToSealed but accepts an io.Writer.
```

### CODE REVIEW

File: token/invocation/schema.go [hunk 1]
```diff
@@ -46,26 +48,36 @@ var _ envelope.Tokener = (*tokenPayloadModel)(nil)
...
+	Aud *string
 
-	// The Command to eventually invoke
+	// The Command
 	Cmd string
-
-	// A unique, random nonce
-	Nonce []byte
+	// The Command's Arguments
+	Args struct {
```
Comment (MichaelMure) [inline token/invocation/schema.go:61]: smells like that could have the Meta treatment, with a dedicated struct

Comment (smoyer64) [reply token/invocation/schema.go:61]: It is EXACTLY like `meta.Meta` and the entire topic of conversation on the "Generalize `meta.Meta` 🧵 "

File: token/invocation/invocation.go [hunk 2]
```diff
@@ -13,73 +13,138 @@ import (
...
+// If no nonce is provided, a random 12-byte nonce is generated. Use the
+// WithNonce or WithEmptyNonce options to specify provide your own nonce
+// or to leave the nonce empty respectively.
+//
+// If no invokedAt is provided, the current time is used. Use the
+// WithInvokedAt or WithInvokedAtIn options to specify a different time.
+//
+// With the exception of the WithMeta option, all other will overwrite
+// the previous contents of their target field.
+func New(iss, sub did.DID, cmd command.Command, prf []cid.Cid, opts ...Option) (*Token, error) {
```
Comment (MichaelMure) [inline token/invocation/invocation.go:64]: I'd have the proofs as options, as there is different ways to pass them: CIDs, tokens, maybe even container

Comment (smoyer64) [reply token/invocation/invocation.go:64]: I thought about that but started with the required arguments being non-option arguments - either way, a little bit of function overloading would help ;)

Comment (smoyer64) [reply token/invocation/invocation.go:65]: The invocation requires this list of `cid.Cid`s that reference the delegations.  Are you talking about providing a list of delegation Tokens or a container providing the delegation Tokens and generatng/extracting the needed `cid.Cid`s from those delegation Tokens?

**Review by fabiobozzo — APPROVED** [review]

**Pull request merged**